pymc_experimental/statespace/core/statespace.py

```diff
@@ -0,0 +1,441 @@
+
+        if filter_type.lower() not in FILTER_FACTORY.keys():
+            raise NotImplementedError(
+                "The following are valid filter types: " + ", ".join(list(FILTER_FACTORY.keys()))
+            )
+
+        if filter_type == "single" and self.k_endog > 1:
+            raise ValueError('Cannot use filter_type = "single" with multiple observed time series')
+
+        self.kalman_filter = FILTER_FACTORY[filter_type.lower()]()
+        self.kalman_smoother = KalmanSmoother()
+
+        if verbose:
+            # TODO: Ask Ricardo how to log this to console without a print
+            # print(
+            #     "Model successfully initialized! The following parameters should be assigned priors inside a PyMC "
+            #     f'model block: {", ".join(self.param_names)}'
+            # )
+            pass
+
+    def unpack_statespace(self):
+        a0 = self.ssm["initial_state"]
+        P0 = self.ssm["initial_state_cov"]
+        T = self.ssm["transition"]
+        Z = self.ssm["design"]
+        R = self.ssm["selection"]
+        H = self.ssm["obs_cov"]
+        Q = self.ssm["state_cov"]
+
+        return a0, P0, T, Z, R, H, Q
+
+    @property
+    def param_names(self) -> List[str]:
+        raise NotImplementedError
+
+    def update(self, theta: at.TensorVariable) -> None:
+        """
+        Put parameter values from vector theta into the correct positions in the state space matrices.
+
+        Parameters
+        ----------
+        theta: TensorVariable
+            Vector of all variables in the state space model
+        """
+        raise NotImplementedError
+
+    def gather_required_random_variables(self) -> at.TensorVariable:
+        """
+        Iterates through random variables in the model on the context stack, matches their names with the statespace
+        model's named parameters, and returns a single vector of parameters to pass to the update method.
+
+        Important point is that the *order of the variables matters*. Update will expect that the theta vector will be
+        organized as variables are listed in param_names. This could be improved...
+
+        Returns
+        ----------
+        theta: TensorVariable
+            A p x 1 theano vector containing all parameters to be estimated among all state space matrices in the
+            system.
+        """
+
+        theta = []
+        pymc_model = modelcontext(None)
+        found_params = []
+        with pymc_model:
+            for param_name in self.param_names:
+                for param in pymc_model.rvs_to_values:
+                    if param.name == param_name:
+                        theta.append(param.ravel())
+                        found_params.append(param.name)
+                for param in pymc_model.deterministics:
+                    if param.name == param_name:
+                        theta.append(param.ravel())
+                        found_params.append(param.name)
```

Comment:
To get a variable: `param = pymc_model["param_name"]`, then check it's a free/observed RV depending on the constraints `assert `param in pymc_model.basic_RVs`